neqo-http3/src/control_stream_local.rs

```diff
@@ -0,0 +1,57 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use crate::hframe::HFrame;
+use crate::Res;
+use neqo_common::{qdebug, Encoder};
+use neqo_transport::{Connection, StreamType};
+
+pub const HTTP3_UNI_STREAM_TYPE_CONTROL: u64 = 0x0;
+
+// The local control stream, responsible for encoding frames and sending them
+#[derive(Default, Debug)]
+pub struct ControlStreamLocal {
+    stream_id: Option<u64>,
+    buf: Vec<u8>,
+}
+
+impl ::std::fmt::Display for ControlStreamLocal {
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        write!(f, "Local control stream {:?}", self.stream_id)
+    }
+}
+
+impl ControlStreamLocal {
+    pub fn send_frame(&mut self, f: HFrame) {
+        let mut enc = Encoder::default();
+        f.encode(&mut enc);
+        self.buf.append(&mut enc.into());
+    }
+    pub fn send(&mut self, conn: &mut Connection) -> Res<()> {
+        if let Some(stream_id) = self.stream_id {
+            if !self.buf.is_empty() {
+                let sent = conn.stream_send(stream_id, &self.buf[..])?;
+                if sent == self.buf.len() {
+                    self.buf.clear();
+                } else {
+                    let b = self.buf.split_off(sent);
+                    self.buf = b;
+                }
+            }
+            return Ok(());
+        }
+        Ok(())
+    }
+
+    pub fn create_control_stream(&mut self, conn: &mut Connection) -> Res<()> {
```

Comment:
Maybe just name `create()` or even `new()`? Also, consider making this a constructor (return `Res<ControlStreamLocal>`). stream_id would no longer be an Option), but instead make ControlStreamLocal at the Http3Connection level an `Option<ControlStreamLocal>`?

EDIT: I see this is just carried over from existing code so this could possibly be addressed separately if desired.

Reply:
The very first version of http3 code has used your suggestion but the code in Http3Connection gets complicated. Also for the control_stream_remote even more complicated because of not being able to borrow 'self' mutable twice, for example here:
  
```
        } else if self
            .control_stream_remote
            .receive_if_this_stream(&mut self.conn, stream_id)?
        {
            qdebug!(
                [self]
                "The remote control stream ({}) is readable.",
                stream_id
            );
            while self.control_stream_remote.frame_reader_done()
                || self.control_stream_remote.recvd_fin()
            {
                self.handle_control_frame()?;
                self.control_stream_remote
                    .receive_if_this_stream(&mut self.conn, stream_id)?;
            }
        }
```